addons/account_asset/account_asset.py

```diff
@@ -1,351 +1,325 @@
+        self.write({'state': 'close'})
+
+    @api.multi
+    def set_to_draft(self):
+        self.write({'state': 'draft'})
+
+    @api.one
+    @api.depends('depreciation_line_ids.move_check')
+    def _amount_residual(self):
+        total_amount = 0.0
+        for line in self.depreciation_line_ids:
+            if line.move_check:
+                total_amount += line.amount
+        self.value_residual = self.value - total_amount - self.salvage_value
+
+    @api.onchange('company_id')
+    def onchange_company_id(self):
+        if self.company_id.currency_id.company_id and self.company_id.currency_id.company_id.id != self.company_id.id:
+            self.currency_id = False
+        else:
+            self.currency_id = self.company_id.currency_id.id
+
+    @api.onchange('value')
+    def onchange_purchase_salvage_value(self):
+        self.value_residual = self.value - self.salvage_value
+
+    @api.one
+    @api.depends('account_move_line_ids')
+    def _entry_count(self):
+        self.entry_count = self.env['account.move.line'].search_count([('asset_id', 'in', self.ids)])
+
+    @api.one
+    @api.constrains('prorata', 'method_time')
+    def _check_prorata(self):
+        if self.prorata and self.method_time != 'number':
+            raise ValidationError(_('Prorata temporis can be applied only for time method "number of depreciations".'))
+
+    @api.onchange('category_id')
+    def onchange_category_id(self):
+        vals = self.onchange_category_id_values(self.category_id.id)
+        # We cannot use 'write' on an object that doesn't exist yet
+        if vals:
+            for k, v in vals['value'].iteritems():
+                setattr(self, k, v)
+
+    def onchange_category_id_values(self, category_id):
         if category_id:
-            category_obj = asset_categ_obj.browse(cr, uid, category_id, context=context)
-            res['value'] = {
-                            'method': category_obj.method,
-                            'method_number': category_obj.method_number,
-                            'method_time': category_obj.method_time,
-                            'method_period': category_obj.method_period,
-                            'method_progress_factor': category_obj.method_progress_factor,
-                            'method_end': category_obj.method_end,
-                            'prorata': category_obj.prorata,
+            category = self.env['account.asset.category'].browse(category_id)
+            return {
+                'value': {
+                    'method': category.method,
+                    'method_number': category.method_number,
+                    'method_time': category.method_time,
+                    'method_period': category.method_period,
+                    'method_progress_factor': category.method_progress_factor,
+                    'method_end': category.method_end,
+                    'prorata': category.prorata,
+                }
```

Comment:
I did it like that for the following reason : in account_asset_invoice.py, the method asset_create needed the values of the onchange method (compatibility between new & old API).